### PR TITLE
Gh 675 create spark import to accumulo from pair rdd operation

### PR DESCRIPTION
**Pull request by james010101101**

### CODE REVIEW

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportJavaRDDOfElementsHandler.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            throw new OperationException("Option outputPath must be set for this option to be run against the accumulostore");
+        }
+        String failurePath = operation.getOption(FAILURE_PATH);
+        if (null == failurePath || failurePath.isEmpty()) {
+            throw new OperationException("Option failurePath must be set for this option to be run against the accumulostore");
+        }
+        Broadcast<AccumuloElementConverter> broadcast = operation.getJavaSparkContext().broadcast(store.getKeyPackage().getKeyConverter());
+        ElementConverterFunction func = new ElementConverterFunction(broadcast);
+        JavaPairRDD<Key, Value> rdd = operation.getInput().flatMapToPair(func);
+        ImportKeyValueJavaPairRDDToAccumulo op = new ImportKeyValueJavaPairRDDToAccumulo.Builder().input(rdd).failurePath(failurePath).outputPath(outputPath).build();
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportJavaRDDOfElementsHandler.java]: Above 4 lines should be final.

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandler.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+public class ImportKeyValueJavaPairRDDToAccumuloHandler implements OperationHandler<ImportKeyValueJavaPairRDDToAccumulo, Void> {
+
+    @Override
+    public Void doOperation(final ImportKeyValueJavaPairRDDToAccumulo operation, final Context context, final Store store) throws OperationException {
+        doOperation(operation, context, (AccumuloStore) store);
+        return null;
+    }
+
+    public void doOperation(final ImportKeyValueJavaPairRDDToAccumulo operation, final Context context, final AccumuloStore store) throws OperationException {
+        AccumuloKeyRangePartitioner partitioner = new AccumuloKeyRangePartitioner(store);
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandler.java]: Should be final.

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandler.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        JavaPairRDD<Key, Value> rdd = operation.getInput();
+        rdd = rdd.repartitionAndSortWithinPartitions(partitioner);
+        rdd.saveAsNewAPIHadoopFile(operation.getOutputPath(), Key.class, Value.class, AccumuloFileOutputFormat.class, getConfiguration(operation));
+        ImportAccumuloKeyValueFiles importAccumuloKeyValueFiles = new ImportAccumuloKeyValueFiles.Builder().inputPath(operation.getOutputPath()).failurePath(operation.getFailurePath()).build();
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandler.java]: Should be final.

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportKeyValuePairRDDToAccumuloHandler.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        return null;
+    }
+
+    private void doOperation(final ImportKeyValuePairRDDToAccumulo operation, final Context context, final AccumuloStore store) throws OperationException {
+        AccumuloKeyRangePartitioner partitioner = new AccumuloKeyRangePartitioner(store);
+        OrderedRDDFunctions orderedRDDFunctions = new OrderedRDDFunctions(operation.getInput(), ORDERING_CLASS_TAG, KEY_CLASS_TAG, VALUE_CLASS_TAG, scala.reflect.ClassTag$.MODULE$.apply(Tuple2.class));
+        PairRDDFunctions pairRDDFunctions = new PairRDDFunctions(orderedRDDFunctions.repartitionAndSortWithinPartitions(partitioner), KEY_CLASS_TAG, VALUE_CLASS_TAG, ORDERING_CLASS_TAG);
+        pairRDDFunctions.saveAsNewAPIHadoopFile(operation.getOutputPath(), Key.class, Value.class, AccumuloFileOutputFormat.class, getConfiguration(operation));
+        ImportAccumuloKeyValueFiles importAccumuloKeyValueFiles = new ImportAccumuloKeyValueFiles.Builder().inputPath(operation.getOutputPath()).failurePath(operation.getFailurePath()).build();
+        store.execute(importAccumuloKeyValueFiles, context.getUser());
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportKeyValuePairRDDToAccumuloHandler.java]: 4 variables above should be final.

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportKeyValuePairRDDToAccumuloHandler.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    protected Configuration getConfiguration(final Operation operation) throws OperationException {
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportKeyValuePairRDDToAccumuloHandler.java]: Seems to be a repeat of code from ImportKeyValueJavaPairRDDToAccumuloHandler ? Can it be moved to a utils class?

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportRDDOfElementsHandler.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        if (null == outputPath || outputPath.isEmpty()) {
+            throw new OperationException("Option outputPath must be set for this option to be run against the accumulostore");
+        }
+        String failurePath = operation.getOption(FAILURE_PATH);
+        if (null == failurePath || failurePath.isEmpty()) {
+            throw new OperationException("Option failurePath must be set for this option to be run against the accumulostore");
+        }
+        ElementConverterFunction func = new ElementConverterFunction(operation.getSparkContext().broadcast(store.getKeyPackage().getKeyConverter(), ACCUMULO_ELEMENT_CONVERTER_CLASS_TAG));
+        RDD<Tuple2<Key, Value>> rdd = operation.getInput().flatMap(func, TUPLE2_CLASS_TAG);
+        ImportKeyValuePairRDDToAccumulo op = new ImportKeyValuePairRDDToAccumulo.Builder().input(rdd).failurePath(failurePath).outputPath(outputPath).build();
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportRDDOfElementsHandler.java]: 3 variables above should be final.

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/AccumuloKeyRangePartitioner.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        Connector connector = null;
+        try {
+            connector = store.getConnection();
+        } catch (StoreException e) {
+            throw new OperationException("Failed to create accumulo connection", e);
+        }
+        String table = store.getProperties().getTable();
+        try {
+            Collection<Text> splits = connector.tableOperations().listSplits(table);
+            String[] arr = new String[splits.size()];
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/AccumuloKeyRangePartitioner.java]: Above 2 lines should be final.

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/AccumuloKeyRangePartitioner.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            return splits.parallelStream().map(text -> text.toString()).collect(Collectors.toList()).toArray(arr);
+        } catch (TableNotFoundException | AccumuloSecurityException | AccumuloException e) {
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/AccumuloKeyRangePartitioner.java]: Add final

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/java/ElementConverterFunction.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        List<Tuple2<Key, Value>> tuples = new ArrayList<>(2);
+        Pair<Key> keys = converterBroadcast.value().getKeysFromElement(e);
+        Value value = converterBroadcast.value().getValueFromElement(e);
+        tuples.add(new Tuple2<>(keys.getFirst(), value));
+        Key second = keys.getSecond();
+        if (second != null) {
+            tuples.add(new Tuple2<>(second, value));
+        }
+        Iterator iter = tuples.listIterator();
+        return iter;
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/java/ElementConverterFunction.java]: Replace with `return tuples.listIterator()`

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/java/ElementConverterFunction.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+       this.converterBroadcast = converterBroadcast;
+    }
+
+    @Override
+    public Iterator<Tuple2<Key, Value>> call(final Element e) throws Exception {
+        List<Tuple2<Key, Value>> tuples = new ArrayList<>(2);
+        Pair<Key> keys = converterBroadcast.value().getKeysFromElement(e);
+        Value value = converterBroadcast.value().getValueFromElement(e);
+        tuples.add(new Tuple2<>(keys.getFirst(), value));
+        Key second = keys.getSecond();
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/java/ElementConverterFunction.java]: Few lines above should be final.

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/scala/ElementConverterFunction.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    private Broadcast<AccumuloElementConverter> converterBroadcast;
+
+    public ElementConverterFunction(final Broadcast<AccumuloElementConverter> converterBroadcast) {
+        this.converterBroadcast = converterBroadcast;
+    }
+
+    @Override
+    public TraversableOnce<Tuple2<Key, Value>> apply(final Element element) {
+        ArrayBuffer<Tuple2<Key, Value>> buf = new ArrayBuffer<>();
+        Pair<Key> keys = new Pair<>();
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/scala/ElementConverterFunction.java]: 2 finals above.

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/scala/ElementConverterFunction.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        Value value = null;
+        try {
+            keys = converterBroadcast.value().getKeysFromElement(element);
+            value = converterBroadcast.value().getValueFromElement(element);
+        } catch (AccumuloElementConversionException e) {
+            LOGGER.error(e.getMessage(), e);
+        }
+        Key first = keys.getFirst();
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/scala/ElementConverterFunction.java]: final

File: library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/scala/ElementConverterFunction.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        if (first != null) {
+            buf.$plus$eq(new Tuple2<>(first, value));
+        }
+        Key second = keys.getSecond();
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/utils/scala/ElementConverterFunction.java]: final

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportJavaRDDOfElementsHandlerTest.java [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+        final JavaSparkContext sparkContext = new JavaSparkContext(sparkConf);
+
+        // Create Hadoop configuration and serialise to a string
+        final Configuration configuration = new Configuration();
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        configuration.write(new DataOutputStream(baos));
+        final String configurationString = new String(baos.toByteArray(), CommonConstants.UTF_8);
+        String outputPath = this.getClass().getResource("/").getPath().toString() + "load";
+        String failurePath = this.getClass().getResource("/").getPath().toString() + "failure";
+        File file = new File(outputPath);
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportJavaRDDOfElementsHandlerTest.java]: Above 3 lines should be final.

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportJavaRDDOfElementsHandlerTest.java [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+        if(file.exists()) {
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportJavaRDDOfElementsHandlerTest.java]: Need space after if

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportJavaRDDOfElementsHandlerTest.java [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+            FileUtils.forceDelete(file);
+        }
+        JavaRDD<Element> elementJavaRDD = sparkContext.parallelize(elements);
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportJavaRDDOfElementsHandlerTest.java]: Add final

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandlerTest.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+                .set("spark.driver.allowMultipleContexts", "true");
+        final JavaSparkContext sparkContext = new JavaSparkContext(sparkConf);
+
+        // Create Hadoop configuration and serialise to a string
+        final Configuration configuration = new Configuration();
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        configuration.write(new DataOutputStream(baos));
+        final String configurationString = new String(baos.toByteArray(), CommonConstants.UTF_8);
+        String outputPath = this.getClass().getResource("/").getPath().toString() + "load";
+        String failurePath = this.getClass().getResource("/").getPath().toString() + "failure";        File file = new File(outputPath);
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandlerTest.java]: Above 2 lines should be final.

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandlerTest.java [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+        if(file.exists()) {
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandlerTest.java]: Space after if

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandlerTest.java [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+            FileUtils.forceDelete(file);
+        }
+        ElementConverterFunction func = new ElementConverterFunction(sparkContext.broadcast(new ByteEntityAccumuloElementConverter(graph1.getSchema())));
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandlerTest.java]: final

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandlerTest.java [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+        JavaPairRDD<Key, Value> elementJavaRDD = sparkContext.parallelize(elements).flatMapToPair(func);
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/javardd/ImportKeyValueJavaPairRDDToAccumuloHandlerTest.java]: final

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportKeyValuePairRDDToAccumuloHandlerTest.java [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        configuration.write(new DataOutputStream(baos));
+        final String configurationString = new String(baos.toByteArray(), CommonConstants.UTF_8);
+        String outputPath = this.getClass().getResource("/").getPath().toString() + "load";
+        String failurePath = this.getClass().getResource("/").getPath().toString() + "failure";
+        File file = new File(outputPath);
+        if(file.exists()) {
+            FileUtils.forceDelete(file);
+        }
+        ElementConverterFunction func = new ElementConverterFunction(sparkContext.broadcast(new ByteEntityAccumuloElementConverter(graph1.getSchema()), ACCUMULO_ELEMENT_CONVERTER_CLASS_TAG));
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportKeyValuePairRDDToAccumuloHandlerTest.java]: 4 lines above should be final.

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportKeyValuePairRDDToAccumuloHandlerTest.java [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+        RDD<Tuple2<Key, Value>> elementRDD = sparkContext.parallelize(elements, 1, ELEMENT_CLASS_TAG).flatMap(func, TUPLE2_CLASS_TAG);
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportKeyValuePairRDDToAccumuloHandlerTest.java]: Add final

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportKeyValuePairRDDToAccumuloHandlerTest.java [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+                .sparkContext(sparkContext)
+                .option(AbstractGetRDDHandler.HADOOP_CONFIGURATION_KEY, configurationString)
+                .build();
+
+        final RDD<Element> rdd = graph1.execute(rddQuery, user);
+        if (rdd == null) {
+            fail("No RDD returned");
+        }
+        Set<Element> results = new HashSet<>();
+        Element[] returnedElements = (Element[]) rdd.collect();
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportKeyValuePairRDDToAccumuloHandlerTest.java]: 2 finals

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportRDDOfElementsHandlerTest.java [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+        final SparkContext sparkContext = new SparkContext(sparkConf);
+
+        // Create Hadoop configuration and serialise to a string
+        final Configuration configuration = new Configuration();
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        configuration.write(new DataOutputStream(baos));
+        final String configurationString = new String(baos.toByteArray(), CommonConstants.UTF_8);
+        String outputPath = this.getClass().getResource("/").getPath().toString() + "load";
+        String failurePath =this.getClass().getResource("/").getPath().toString() + "failure";
+        File file = new File(outputPath);
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportRDDOfElementsHandlerTest.java]: 3 finals above

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportRDDOfElementsHandlerTest.java [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+                .sparkContext(sparkContext)
+                .option(AbstractGetRDDHandler.HADOOP_CONFIGURATION_KEY, configurationString)
+                .build();
+
+        final RDD<Element> rdd = graph1.execute(rddQuery, user);
+        if (rdd == null) {
+            fail("No RDD returned");
+        }
+        Set<Element> results = new HashSet<>();
+        Element[] returnedElements = (Element[]) rdd.collect();
```
Comment (gaffer01) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/scalardd/ImportRDDOfElementsHandlerTest.java]: 2 finals

File: library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/operation/javardd/ImportJavaRDDOfElements.java [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import uk.gov.gchq.gaffer.data.element.Element;
+import uk.gov.gchq.gaffer.operation.AbstractOperation;
+import uk.gov.gchq.gaffer.operation.VoidOutput;
+import uk.gov.gchq.gaffer.operation.serialisation.TypeReferenceImpl;
+
+public class ImportJavaRDDOfElements extends AbstractOperation<JavaRDD<Element>, Void>implements VoidOutput<JavaRDD<Element>> {
```
Comment (gaffer01) [inline library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/operation/javardd/ImportJavaRDDOfElements.java]: Space before implements

File: library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/operation/javardd/ImportJavaRDDOfElements.java [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+    @Override
+    protected TypeReference createOutputTypeReference() {
+        return new TypeReferenceImpl.Void();
+    }
+
+    protected abstract static class BaseBuilder<CHILD_CLASS extends BaseBuilder<?>>
+            extends AbstractOperation.BaseBuilder<ImportJavaRDDOfElements, JavaRDD<Element>, Void, CHILD_CLASS> {
+
+        public BaseBuilder() {
+
```
Comment (gaffer01) [inline library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/operation/javardd/ImportJavaRDDOfElements.java]: Delete empty line

File: library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/operation/javardd/ImportJavaRDDOfElements.java [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+            super(new ImportJavaRDDOfElements());
+        }
+
```
Comment (gaffer01) [inline library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/operation/javardd/ImportJavaRDDOfElements.java]: Delete empty line

File: library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/operation/scalardd/ImportRDDOfElements.java [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+        return new TypeReferenceImpl.Void();
+    }
+
+    protected abstract static class BaseBuilder<CHILD_CLASS extends BaseBuilder<?>>
+            extends AbstractOperation.BaseBuilder<ImportRDDOfElements, RDD<Element>, Void, CHILD_CLASS> {
+
+        public BaseBuilder() {
+            super(new ImportRDDOfElements());
+        }
+
```
Comment (gaffer01) [inline library/spark/spark-library/src/main/java/uk/gov/gchq/gaffer/spark/operation/scalardd/ImportRDDOfElements.java]: Delete empty line

**Comment by p013570** [comment]
I'll implement the requested changes.

**Comment by p013570** [comment]
All the comments have been addressed. @gaffer01 please could you check over the changes - thanks.

**Pull request merged**